flask-version/metagenomongo/app.py

```diff
@@ -90,19 +91,21 @@ def save_file_server(output_value,file_name):
         with open(filepath, 'w', encoding='utf-8') as f:
             f.write(output_value)
     except:
-        print()
+        print("write")
     remote_path = os.getenv('META_REMOTE_PATH')
     key_path = os.getenv('META_KEY_PATH')
-    # print(f"key_path:{key_path}")
-    # print(f"remote_path:{remote_path}")
-    scp_command = ['scp', '-i', key_path, filepath, remote_path]
+    if key_path == None:
+        print(f"Set META_KEY_PATH")
+        return
     try:
+        scp_command = ['scp', '-i', '../'+key_path, filepath, remote_path]
```

Comment:
@iwmstjp What are you trying to achieve with adding `..` to the key path? This doesn't seem to make sense to me. The user provides the key path on the command line. If they provide an absolute path, like `/home/ubuntu/.ssh/keyfile`, then the application won't find the file. In addition, the value of the environment variable is independent of which directory the user is starting the app from once it's set, it's just a string, while `..` is relative to what the Python interpreter sees as the current working directory. It isn't self-evident what the latter is. Is it the location of `app.py` or the location of the directory from which the app is started if it's started from a different directory using a relative path? The user will have not idea. It isn't clear why the key file should be in the `..` directory at all. In addition, if the keyfile can't be found, the user gets no feedback from the app why it can't be found, i.e. where the app is exactly looking for the keyfile.

To summarise, this seems like a very bad design decision to me for lots of reasons. Sourcery's suggestion doesn't help much. Why should the path of the key file be relative to the parent directory of the directory that `app.py` is in? This does not help at all.

On lines 97-99 you print a message telling the user to set a key path if it has not been set, but this is quite pointless. If the key path is not set, the try-except statement from on lines 100 to 108 will fail anyway, but it will also fail if the path is set incorrectly. The user will get no feedback about the fact that the upload failed. These are also bad design decisions.

Revise the code as follows:
- Check whether the necessary environment variables have been set when the application starts up, not when `save_file_server` is called. Do it after the part when the secret key is checked. The application should try to open the key file for reading. If it can't be opened, let the application crash with a FileNotFoundError, stating the path where the key file was not found.
- Similarly check whether `META_REMOTE_PATH` has been set, and if not, make the startup crash with a ValueError similar to the secret key.

State in the readme that the path to the keyfile should be specified as an absolute path.

As I mentioned in the meeting, also explain in the readme what the remote path variable is for and that it's a remote machine that can be accessed by ssh with an ssh key. You should probably try generating the usage guide for the whole application using ChatGPT/Claude/Gemini by attaching the files of the application as context for them. It's worth a try. If it works, it could save you plenty of time.

Reply:
I modified this part. Key_path is an absolute path now.
scp_command = ['scp', '-i', key_path, filepath, remote_path]
I added checking for the necessary environment variables in the beggining of the code.